Completely Remove FutureMessage from RPC agents

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50236 Apply clang-format to rpc cpp files
* #50029 Completely remove FutureMessage type
* **#50028 Completely Remove FutureMessage from RPC agents**
* #50027 Completely Remove FutureMessage from RPC cpp tests
* #50026 Remove FutureMessage from RPC request callback logic
* #50025 Completely remove FutureMessage from FaultyProcessGroupAgent
* #50024 Remove FutureMessage from sender TensorPipeAgent
* #50023 Remove FutureMessage from sender ProcessGroupAgent
* #50020 Completely remove FutureMessage from distributed autograd
* #50005 Completely remove FutureMessage from RPC TorchScript implementations
* #50004 Completely remove FutureMessage from RRef Implementations
* #49995 Replace FutureMessage with ivalue::Future in RpcAgent retry logic
* #49960 Replace FutureMessage with ivalue::Future in RRefContext
* #49927 Replace FutureMessage with ivalue::Future in distributed/autograd/utils.*
* #49906 Let RpcAgent::send() return JitFuture

Differential Revision: [D25753887](https://our.internmc.facebook.com/intern/diff/D25753887)

mrshenli committed:
Completely Remove FutureMessage from RPC agents

[ghstack-poisoned]

mrshenli committed:
Update on "Completely Remove FutureMessage from RPC agents"


Differential Revision: [D25753887](https://our.internmc.facebook.com/intern/diff/D25753887)

[ghstack-poisoned]

lw commented on torch/csrc/distributed/rpc/process_group_agent.cpp:
```diff
@@ -514,22 +514,25 @@ bool ProcessGroupAgent::handleRecv(RecvWork& work) {
   if (message.isRequest()) {
     ++serverActiveCalls_;
-    std::shared_ptr<FutureMessage> futureResponse;
+    auto futureResponse =
+        std::make_shared<JitFuture>(at::AnyClassType::get());
```
These lines are different than the ones they're replacing: before we were creating an empty (null) shared_ptr, here we create a "full" shared_ptr that points to an uncompleted future. It also appears that immediately after we re-set this shared_ptr to a new value, hence destroying that uncompleted future. I think it makes sense to avoid creating it in the first place.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -477,17 +477,14 @@ void TensorPipeAgent::sendCompletedResponseMessage(
-  if (!error) {
+  if (!futureResponseMessage->hasError()) {
+    Message&& responseMessage =
+        std::move(*futureResponseMessage->value().toCustomClass<Message>());
```
I wonder: does it make sense to add a moveValue() method to ivalue::Future? For cases like this where we don't care about the future anymore and just want its content?

(Personally I think the "idiomatic" way would be to have a &&-qualified value() method but I've seen that in PyTorch that's not always the convention...)

mrshenli replied:
Sure, I can add that. cc @wanchaol if there are concerns for adding `value() &&`.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -477,17 +477,14 @@ void TensorPipeAgent::sendCompletedResponseMessage(
-  const c10::optional<utils::FutureError> error =
-      futureResponseMessage->error();
-  Message&& responseMessage = std::move(*futureResponseMessage).moveValue();
-  responseMessage.setId(messageId);
```
Where did this line go? Is it not needed?

mrshenli replied:
Yep, I don't think this is necessary as request cbs have already set the id for the response message. But let me add it back for safety.

lw commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -572,12 +570,14 @@ void TensorPipeAgent::respond(std::shared_ptr<tensorpipe::Pipe>& pipe) {
 
-          std::shared_ptr<FutureMessage> futureResponseMessage;
+          auto futureResponseMessage =
+              std::make_shared<JitFuture>(at::AnyClassType::get());
```
Same as above: we're allocating an uncompleted JitFuture just to immediately replacing it (and thus deallocating it). It's probably better to leave the shared_ptr unset.

mrshenli committed:
Update on "Completely Remove FutureMessage from RPC agents"


Differential Revision: [D25753887](https://our.internmc.facebook.com/intern/diff/D25753887)

[ghstack-poisoned]

mrshenli committed:
Update on "Completely Remove FutureMessage from RPC agents"


Differential Revision: [D25753887](https://our.internmc.facebook.com/intern/diff/D25753887)

[ghstack-poisoned]
